src/core/lib/security/credentials/tls/grpc_tls_certificate_distributor.cc

```diff
@@ -0,0 +1,290 @@
+    if (it == this->pem_root_certs_.end()) {
+      this->pem_root_certs_.insert(
+          std::make_pair(root_cert_name, std::string(*pem_root_certs)));
+    } else {
+      it->second = std::string(*pem_root_certs);
+    }
+    updated_root_cert_name = absl::make_optional(root_cert_name);
+  }
+  if (pem_key_cert_pairs) {
+    const auto& it = this->pem_key_cert_pair_.find(identity_cert_name);
+    if (it == this->pem_key_cert_pair_.end()) {
+      this->pem_key_cert_pair_.insert(
+          std::make_pair(identity_cert_name, *pem_key_cert_pairs));
+    } else {
+      it->second = *pem_key_cert_pairs;
+    }
+    updated_identity_cert_name = absl::make_optional(identity_cert_name);
+  }
+  if (updated_root_cert_name || updated_identity_cert_name) {
+    this->CertificatesUpdated(updated_root_cert_name,
+                              updated_identity_cert_name);
+  }
+}
+
+void grpc_tls_certificate_distributor::SetRootCerts(
+    std::string root_cert_name, absl::string_view pem_root_certs) {
+  // move the root_cert_name to avoid unnecessary copy.
+  SetKeyMaterials(std::move(root_cert_name),
+                  absl::make_optional(pem_root_certs), "", absl::nullopt);
+};
+
+void grpc_tls_certificate_distributor::SetKeyCertPairs(
+    std::string identity_cert_name, PemKeyCertPairList pem_key_cert_pairs) {
+  // move the identity_cert_name and pem_key_cert_pairs to avoid unnecessary
+  // copies.
+  SetKeyMaterials("", absl::nullopt, std::move(identity_cert_name),
+                  absl::make_optional(std::move(pem_key_cert_pairs)));
+};
+
+bool grpc_tls_certificate_distributor::HasRootCerts(
+    const std::string& root_cert_name) {
+  grpc_core::MutexLock lock(&mu_);
+  const auto& it = this->pem_root_certs_.find(root_cert_name);
+  return it != this->pem_root_certs_.end();
+};
+
+bool grpc_tls_certificate_distributor::HasKeyCertPairs(
+    const std::string& identity_cert_name) {
+  grpc_core::MutexLock lock(&mu_);
+  const auto& it = this->pem_key_cert_pair_.find(identity_cert_name);
+  return it != this->pem_key_cert_pair_.end();
+};
+
+void grpc_tls_certificate_distributor::WatchTlsCertificates(
+    std::unique_ptr<TlsCertificatesWatcherInterface> watcher,
+    absl::optional<std::string> root_cert_name,
+    absl::optional<std::string> identity_cert_name) {
+  grpc_core::MutexLock lock(&mu_);
+  TlsCertificatesWatcherInterface* watcher_ptr = watcher.get();
+  if (watcher_ptr == nullptr) {
+    return;
+  }
+  WatcherInfo info = {std::move(watcher), root_cert_name, identity_cert_name};
+  watchers_.insert(std::make_pair(watcher_ptr, std::move(info)));
+  if (root_cert_name) {
+    const auto& it = this->watch_status_.find(*root_cert_name);
+    // We will only notify the caller(e.g. the Producer) the first time a
+    // particular certificate name is being watched.
+    if (it == this->watch_status_.end() ||
+        it->second.root_cert_watcher_cnt == 0) {
+      bool identity_cert_watched = false;
```

Comment:
echo to my other comments. Separate watch_status_callback into two will be the code much cleaner and easy to understand.